### PR TITLE
feat: allow injection of pgstac connection

### PR DESCRIPTION
**Pull request by ceholden**
## Technical Context
- **Related Issues:** N/A but can create if it helps
- **Breaking Changes:** No - existing code should "just work" since new kwarg is 

## Description

This PR adds a new keyword argument to PgSTAC related functions allowing a user to control the connection details. This should be useful generally (e.g., using a connection pool), but specifically came up in resolving using this code to connect to PgSTAC hosted behind a RDS Proxy:

>FATAL: Feature not supported: RDS Proxy currently doesn't support command-line options

This PR resolves this by giving more control for the connection to users with a small downside of needing more ownership (e.g., setting `search_path`, see `_connect` docstring). This should be backwards compatible.

---

## Checklist
- [x] **Linting:** Code is formatted and linted
- [x] **Tests:** Tests pass. I have included new tests for these changes where applicable.
- [x] **Edge Cases:** I have manually verified "unhappy paths" and edge cases beyond the basic success criteria (e.g., database connection timeouts, malformed input, strict mapping rejections).
- [x] **Documentation:** I have updated `README.md` to reflect any new environment variables, configuration changes, or breaking updates.
- [x] **Accountability:** I can explain the implementation logic for every line of code submitted.

---

## AI tool usage
- [x] AI (Copilot or something similar) supported my development of this PR. See our [policy about AI tool use](https://stac-utils.github.io/ai-contribution-policy). Use of AI tools *must* be indicated.
    - I used Claude to help review this before submitting

---
> **Policy:** We require a "human-in-the-loop." You are the author and are fully accountable for all submitted code. Please ensure all tool-generated content is thoroughly reviewed before submission to ensure it is not an "extractive contribution" that squanders maintainer time.

### CODE REVIEW

File: stac_geoparquet/pgstac_reader.py [hunk 1]
```diff
@@ -128,18 +129,41 @@ def pgstac_dsn(conninfo: str | None, statement_timeout: int | None) -> str:
...
+@contextmanager
+def _connect(
     conninfo: str | None,
+    statement_timeout: int | None,
+    conn_factory: Callable[[], psycopg.Connection] | None,
+) -> Iterator[psycopg.Connection]:
+    """
+    Yield a pgstac connection, via connection string or a caller-supplied `conn_factory`.
+
+    Exactly one of the two must be given. `conn_factory` users are responsible for
```
Comment (gadomski) [inline stac_geoparquet/pgstac_reader.py]: This might be a personal preference, but I find exclusive arguments non-ergonomic. What do you think about making `conninfo` a `str | Callable`?

Comment (ceholden) [reply stac_geoparquet/pgstac_reader.py]: Great idea - much nicer, thanks!

Updated in https://github.com/stac-utils/stac-geoparquet/pull/145/commits/73590b0f76df8b84ce9e49895514dea3e6367212

File: stac_geoparquet/pgstac_reader.py [hunk 2]
```diff
@@ -128,8 +129,27 @@ def pgstac_dsn(conninfo: str | None, statement_timeout: int | None) -> str:
...
+        with conninfo() as conn:
+            yield conn
+        return
+    with psycopg.connect(pgstac_dsn(conninfo, statement_timeout)) as conn:
+        yield conn
+
+
 def pgstac_to_iter(
-    conninfo: str | None,
+    conninfo: str | Callable[[], psycopg.Connection],
```
Comment (gadomski) [inline stac_geoparquet/pgstac_reader.py]: Sorry, I was inprecise with my previous comment — we probably want to keep `None`, as (I believe) this enables loading from environment variables via https://github.com/stac-utils/stac-geoparquet/blob/b5097d30965bb902cf77c34322351e21801ddad6/stac_geoparquet/pgstac_reader.py#L120

```suggestion
    conninfo: str | Callable[[], psycopg.Connection] | None,
```

**Review by gadomski — CHANGES_REQUESTED** [review]

File: stac_geoparquet/pgstac_reader.py (hunk 2 above)
Comment (ceholden) [reply stac_geoparquet/pgstac_reader.py]: Ah thanks for catching that. I updated to have that type signature again in https://github.com/stac-utils/stac-geoparquet/pull/145/commits/267c2c0cba1b8433b1413bb99717afd2dd83f97e

**Review by gadomski — APPROVED** [review]
